pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,228 @@
+	return &RaftTracer{ctx: ctx}
+}
+
+// RegisterRemote is used to register a remote trace. This is used when we
+// receive a raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(
+	ctx context.Context,
+	index kvpb.RaftIndex,
+	traceID tracingpb.TraceID,
+	remoteSpanID tracingpb.SpanID,
+) {
+	if r.numRegistered.Load() > maxTraces {
+		return
+	}
+	r.numRegistered.Add(1)
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	t := newRemoteTrace(r.ctx, traceID, remoteSpanID)
+	t.logf(1, "trace index (i%d) from remote %s", index, tracingID(traceID, remoteSpanID))
+	r.m.Store(index, &t)
+}
+
+// withContext allows us to get the context from the object.
+type withContext interface {
+	Context() context.Context
+}
+
+// Register is called on an entry that is about to be proposed. This will begin
+// logging all subsequent updates to this entry.
+func (r *RaftTracer) Register(baseCtx withContext, ent raftpb.Entry) {
+	if r.numRegistered.Load() > maxTraces {
+		return
+	}
+	r.numRegistered.Add(1)
+	// We store all the entries in the map. This is because we don't know which
+	// one was traced. Typically this is called with only a single entry.
+	t := newLocalTrace(r.ctx, baseCtx)
+	// Log the identical entry to both to make it easier to match up later.
+	t.logf(1, "registering entry (i%d) for tracing", ent.Index)
+	r.m.Store(kvpb.RaftIndex(ent.Index), &t)
+}
+
+// IndexContext is a pair of a raft index and its tracing context.
+type IndexContext struct {
+	Index uint64
+	Ctx   context.Context
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []*kvserverpb.TracedEntry {
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// Close will unregister all the traces and log them. It should be called if the
+// replica is destroyed as it is possible some traces were never closed.
+func (r *RaftTracer) Close() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.numRegistered.Add(-1)
+		r.m.Delete(index)
+		t.logf(5, "cleanup log mark (i%d) during close", index)
+		return true
+	})
+}
+
+// traceIfCovered will log the message if it is covered by a trace. Additionally
+// it returns any saved contexts by index for sending to remote nodes.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []*kvserverpb.TracedEntry {
+	var tracedEntries []*kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		switch m.Type {
+		case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
```

Comment:
This switch-case seems redundant, since the `traceIfCovered` method is called under the exact same condition.

Reply:
Good point! Removed. If we later create different logging for different message types we can bring this back.